tools/spi_nor_config_generator/spi_nor_cfg_gen.py

```diff
@@ -0,0 +1,463 @@
+        if not model_var and not chip_var:
+            # No variations
+            plain_all_entries.append(entry)
+            continue
+        lnames = separate_names(entry["chip_name"], model_var, chip_var)
+        for name in lnames:
+            new_entry = dict()
+            for key in entry:
+                if key == "chip_name":
+                    new_entry[key] = name
+                else:
+                    new_entry[key] = entry[key]
+            plain_all_entries.append(new_entry)
+
+    # Search for duplicates
+    seen_id = set()
+    seen_name = set()
+    seen_vendor = set()
+    uniqs = list()
+    for entry in plain_all_entries:
+        if entry["id"] not in seen_id and entry["chip_name"] not in seen_name:
+            uniqs.append(entry)
+            seen_id.add(entry["id"])
+            seen_name.add(entry["chip_name"])
+        elif entry["chip_name"] not in seen_name:
+            duplicate = False
+            # Check if it has '.' characters so we can treat them as regex
+            if '.' in entry["chip_name"]:
+                pattern = re.compile(entry["chip_name"])
+                for u in uniqs:
+                    result = pattern.match(u["chip_name"])
+                    if result != None:
+                        duplicate = True
+                        break
+
+            if not duplicate:
+                uniqs.append(entry)
+                seen_name.add(entry["chip_name"])
+
+    # Override config if possible
+    for entry in uniqs:
+        # These are just defaults, override them if do not match
+        entry["dummy_bits"] = 8
+        entry["addr_widths"] = [True, False, False, False] # [single, dual, quad, octal]
+        # Override configs
+        if entry["chip_name"] in override_chip_cfg.keys():
+            for key in override_chip_cfg[entry["chip_name"]]:
+                entry[key] = override_chip_cfg[entry["chip_name"]][key]
+        # We can separate device ID from vendor ID here
+        entry["device_id"] = ((entry["id"] >> 24) & 0xFFFF)
+        entry["vendor_id"] = ((entry["id"] >> 40) & 0xFF)
+        del entry["id"]
+
+    # Save JSON
+    if output_json != None:
+        with open(output_json, "w") as json_file:
+            for entry in uniqs:
+                json.dump(entry, json_file)
+
+        os.system('jq -c . %s > %s' % (output_json, output_json + ".pretty"))
+        os.remove(output_json)
+        os.rename(output_json + ".pretty", output_json)
+        logger.std_print("JSON SPI NOR cofigs generated to %s" % (output_json))
+
+    with open(output_class, "w") as pyfile:
+        pyfile.write("""# Generated using 'spi_nor_cfg_gen.py'
+
+""")
+        for entry in uniqs:
+            name = entry["chip_name"].replace('-', '_').replace('.', 'x')
+            # Add 'X' character for chip name starting with digit
+            if entry["chip_name"][0].isdigit():
+                name = 'X' + name
+
+            generated_class = '''class {pn}(object):
```

Comment:
Can we move this generated class to a seperate template file?

Reply:
Function moved [here](https://github.com/litex-hub/litespi/pull/1/files?file-filters%5B%5D=.patch&file-filters%5B%5D=.py&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile&hide-deleted-files=true#diff-0a7a8b224ab31672a326a8755a06eb3f)